### DISCUSSION TITLE
Linkerd Proxy and InternalTrafficPolicy

### DISCUSSION THREAD

**Discussion by aa250816** [linkerd/linkerd2#10130] [Q&A]
For pods with the linkerd sidecar injected linkerd replaces the kube-proxy with it’s own linkerd-proxy that functions similar to the kube-proxy and uses the services and associated endpoints to route traffic. It seems like it does not support or respect the [internalTrafficPolicy](https://kubernetes.io/docs/concepts/services-networking/service-traffic-policy/) of a service. In testing this on a cluster traffic from a pod one node could get routed to an endpoint for a pod on an a different node even with `internalTrafficPolicy: local` on the service used. For example: 

![image](https://user-images.githubusercontent.com/66782813/212179938-0d6a7e64-ff37-4c53-8c09-9c1e9eb4cd2c.png)

So my question is there any way to replicate `internalTrafficPolicy: local` when linkerd is enabled? I’ve come across [topology aware service routing](https://kubernetes.io/docs/concepts/services-networking/service-topology) which I initially stayed away from since it has a lot more complexity which would give us a lot more room to make mistakes in the configuration especially when internalTrafficPolicy gets us what we need. [Linkerd seemed to support it](https://linkerd.io/2020/11/23/topology-aware-service-routing-on-kubernetes-with-linkerd/). The topology aware keys has been deprecated and replaced by [topology aware hints](https://kubernetes.io/docs/concepts/services-networking/topology-aware-hints/) which again linkerd does seem to support but [does not have much if any docs](https://github.com/linkerd/website/issues/1519) on. Reading into topology aware hints further doesn’t seem like it will get us exactly what we want since it has some constraints about when it applies the hints and doesn’t have any guarantees about routing like internal traffic policy.

If there is not any way to replicate `internalTrafficPolicy: local` today are there any plans for the linkerd-proxy to take that into account like it does for the topology aware hints?

**Comment by risingspiral** [comment]
I believe this was introduced as a stable feature in K8s with 1.26. We're currently targeting 1.26 support with the upcoming release of Linkerd 2.13 targeting early this year. We'll have more specifics on a release date and internalTrafficPolicy: local support soon.

cc @adleong

**Comment by yc185050** [comment]
Created a PR to address this issue. https://github.com/linkerd/linkerd2/pull/10186
